BinaryFormatter removal (#42054)
dotnet/docs: docs/core/compatibility/serialization/9.0/binaryformatter-removal.md

---

> title: "Breaking change: In-box BinaryFormatter is unusable"

Reviewer: This sounds more like a problem with the quality of the user experience, rather than a deliberate design decision.

```suggestion
title: "Breaking change: In-box BinaryFormatter implementation was removed and always throws"
```

ms.date: 08/06/2024
---
# In-box BinaryFormatter is unusable

The "in box" <xref:System.Runtime.Serialization.Formatters.Binary.BinaryFormatter> implementation now throws exceptions at run time in all cases. This is the final stage of the [BinaryFormatter obsoletion plan](https://github.com/dotnet/designs/blob/main/accepted/2020/better-obsoletion/binaryformatter-obsoletion.md).

## Previous behavior

You could construct a <xref:System.Runtime.Serialization.Formatters.Binary.BinaryFormatter> instance and use it to serialize and deserialize payloads.

## New behavior

Starting in .NET 9, the in-box <xref:System.Runtime.Serialization.Formatters.Binary.BinaryFormatter> implementation throws exceptions on use, even with the settings that previously enabled its use. Those settings are also removed.

## Version introduced

.NET 9 Preview 6

## Type of breaking change

This change is a [behavioral change](../../categories.md#behavioral-change).

## Reason for change

`BinaryFormatter` is an insecure format and the cause of many security bugs. Removing it from the framework increases the overall safety of .NET.

> ## Recommended action

Reviewer: We should probably link to the BinaryFormatter migration topic, but that's not merged yet (#41564).

If your code uses `BinaryFormatter`, you should select a new serialization format and migrate your code.

> If you judge the risk of `BinaryFormatter` acceptable for your use cases and you're committed to using a class that can't be made secure, you'll still be able to use `BinaryFormatter` through a separate NuGet package. This package is marked as "vulnerable".

Reviewer 2: The package won't be marked as vulnerable. At least not at this time.

```suggestion
If you judge the risk of `BinaryFormatter` acceptable for your use cases and you're committed to using a class that can't be made secure, you'll still be able to use `BinaryFormatter` through a separate unsupported NuGet package.
```

> When more detailed information is available, a link will be added here.

Reviewer 2: We merged the migration guide to main.

```suggestion
For more information including guidance on alternative serializers, refer to the [BinaryFormatter migration guide](/dotnet/standard/serialization/binaryformatter-migration-guide/index.md).
```

> For more information including guidance on alternative serializers, refer to the [BinaryFormatter migration guide](/dotnet/standard/serialization/binaryformatter-migration-guide/index.md).

Author: ```suggestion
For more information, including guidance on alternative serializers, see the [BinaryFormatter migration guide](../../../../standard/serialization/binaryformatter-migration-guide/index.md).
```
